### PR TITLE
executor: Add a limitation about float data type 

### LINKED ISSUE #16039: data type float needs a limitation
## Description
need to add a limitation for float, just as MySQL does:

TiDB v3.0.12:
```
mysql> insert into t1 values(99999999999999999999999999999999999999);
Query OK, 1 row affected (0.04 sec)

mysql> insert into t1 values(999999999999999999999999999999999999999);
Query OK, 1 row affected (0.01 sec)

mysql> insert into t1 values(999999999999999999999999999999999999999999999999999999999999999999999999999999999);
Query OK, 1 row affected (0.01 sec)

mysql> insert into t1 values(9999999999999999999999999999999999999999999999999999999999999999999999999999999999);
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your TiDB version for the right syntax to use line 1 column 104 near "9999999999999999999999999999999999999999999999999999999999999999999999999999999999)"decimal literal: [types:1690]%s value is out of range in '%s' 
mysql> select * from t1;
+------+
| a    |
+------+
| 1e38 |
| +Inf |
| +Inf |
+------+
3 rows in set (0.00 sec)

mysql> desc t1;
+-------+-------+------+------+---------+-------+
| Field | Type  | Null | Key  | Default | Extra |
+-------+-------+------+------+---------+-------+
| a     | float | YES  |      | NULL    |       |
+-------+-------+------+------+---------+-------+
1 row in set (0.00 sec)
```

MySQL 5.7.28:
```
mysql> select version();
+-----------+
| version() |
+-----------+
| 5.7.28    |
+-----------+
1 row in set (0.01 sec)

mysql> select tidb_version();
ERROR 1305 (42000): FUNCTION gin.tidb_version does not exist
mysql> insert into t1 values(99999999999999999999999999999999999999);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t1 values(999999999999999999999999999999999999999);
ERROR 1264 (22003): Out of range value for column 'a' at row 1
mysql> select * from t1;
+------+
| a    |
+------+
| 1e38 |
+------+
1 row in set (0.00 sec)
```

## SIG slack channel
[#sig-exec](https://tidbcommunity.slack.com/archives/CMRD79DRR)

## Score
- 300

## Mentor
 * @lzmhhh123

### PR DESCRIPTION
**Pull request by erwadba**
### What problem does this PR solve?

Issue Number: close #16039 

Problem Summary:

### What is changed and how it works?

What's Changed:
When TypeFloat column value > math.MaxFloat32 in convertToFloat, generate an ErrOverflow error.

How it Works:

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)
```
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v4.0.0-beta.2-1232-g22a858a05-dirty
Edition: Community
Git Commit Hash: 22a858a05f4829e2a94963cc1b3cc66f90ad8224
Git Branch: floatlimit
UTC Build Time: 2020-09-16 02:54:05
GoVersion: go1.13.8
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)

mysql> create table t(id float);
Query OK, 0 rows affected (0.03 sec)

mysql> insert into t values(999999999999999999999999999999999999999);
ERROR 1264 (22003): Out of range value for column 'id' at row 1
mysql> insert into t values(-999999999999999999999999999999999999999);
ERROR 1264 (22003): Out of range value for column 'id' at row 1
mysql> insert ignore into t values(-999999999999999999999999999999999999999);
Query OK, 1 row affected, 1 warning (0.01 sec)

mysql> insert ignore into t values(999999999999999999999999999999999999999);
Query OK, 1 row affected, 1 warning (0.01 sec)

mysql> show warnings;
+---------+------+----------------------------------+
| Level   | Code | Message                          |
+---------+------+----------------------------------+
| Warning | 1690 | %s value is out of range in '%s' |
+---------+------+----------------------------------+
1 row in set (0.00 sec)

mysql> select * from t;
+---------------+
| id            |
+---------------+
| -3.4028235e38 |
|  3.4028235e38 |
+---------------+
2 rows in set (0.00 sec)

mysql>

```
Side effects

n/a

### Release note 

- No release note

### CODE REVIEW

**Comment by erwadba** [comment]
@ti-srebot /run-all-tests

**Comment by erwadba** [comment]
@ti-srebot /rebuild

**Comment by erwadba** [comment]
@SunRunAway PTAL

**Review by zz-jason — COMMENTED** [review]
LGTM

File: table/column.go [hunk 1]
```diff
@@ -191,6 +192,11 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, r
 		truncateTrailingSpaces(&casted)
 	}
 
+	if col.Tp == mysql.TypeFloat && casted.GetFloat32() > math.MaxFloat32 {
```
Comment (lzmhhh123) [inline table/column.go:195]: Why don't we move this logic to function `ConvertTo`?

**Comment by erwadba** [comment]
@ti-srebot /run-all-tests

File: table/column.go (hunk 1 above)
Comment (erwadba) [reply table/column.go]: Thx.

File: types/datum.go [hunk 2]
```diff
@@ -880,7 +880,12 @@ func (d *Datum) convertToFloat(sc *stmtctx.StatementContext, target *FieldType)
 		err = err1
 	}
 	if target.Tp == mysql.TypeFloat {
-		ret.SetFloat32(float32(f))
+		if f > math.MaxFloat32 {
```
Comment (dyzsr) [inline types/datum.go:883]: Do we have to check `f < -math.MaxFloat32` here?

Comment (erwadba) [reply types/datum.go]: Thx.

**Comment by erwadba** [comment]
The bug was already fixed by #20158

**Pull request closed without merging**